stream-chat-android-core/src/main/java/io/getstream/chat/android/client/call/CoroutineCall.kt

```diff
@@ -32,24 +35,23 @@ public class CoroutineCall<T : Any>(
 ) : Call<T> {
 
     private var job: Job? = null
+    private val canceled = AtomicBoolean(false)
 
-    internal suspend fun awaitImpl(): Result<T> {
-        return withContext(scope.coroutineContext) {
-            suspendingTask()
-        }
+    override fun execute(): Result<T> = runBlocking { await() }
+    override suspend fun await(): Result<T> = withContext(scope.coroutineContext) {
+        suspendingTask().takeUnless { canceled.get() } ?: callCanceledError()
     }
 
     override fun cancel() {
+        canceled.set(true)
         job?.cancel()
     }
 
-    override fun execute(): Result<T> {
-        return runBlocking(block = suspendingTask)
-    }
-
     override fun enqueue(callback: Call.Callback<T>) {
+        println("[JcLog] CoroutineCall::enqueue")
         job = scope.launch {
-            val result = suspendingTask()
+            val result = await()
+            yield()
             withContext(DispatcherProvider.Main) {
                 callback.onResult(result)
             }
```

Comment:
Should we check that the call is not canceled, cause we might have switched the thread?
```suggestion
            withContext(DispatcherProvider.Main) {
                if (!canceled.get()) {
                    callback.onResult(result)
                }
            }
```